azurelinuxagent/common/protocol/extensions_config_retriever.py

```diff
@@ -0,0 +1,374 @@
+        self._fabric_changed_detail = None
+        self._reason = None
+        self._reset_if_necessary()
+
+    def get_ext_config(self, incarnation, ext_conf_uri):
+        # If we don't have a uri, return an empty extensions config
+        is_startup = False
+        if ext_conf_uri is None:
+            return GenericExtensionsConfig(ExtensionsConfig(None), False, self)
+
+        # Get the Fabric goal state and whether it changed
+        fabric_changed = self._get_fabric_changed(incarnation)
+
+        # Get the VmArtifactsProfile and whether fast track changed, if enabled
+        artifacts_profile, fast_track_changed = self._get_fast_track_details()
+
+        self._pending_mode = self._decide_what_to_process(fabric_changed, fast_track_changed)
+        if self._last_mode is None:
+            is_startup = True
+            logger.info("Processing first mode {0}. Reason={1}", self._pending_mode, self._reason)
+        elif self._pending_mode != self._last_mode:
+            logger.info("Processing from previous mode {0}. New mode is {1}. Reason={2}",
+                        self._last_mode, self._pending_mode, self._reason)
+
+        extensions_config = None
+        changed = False
+        if self._pending_mode == GOAL_STATE_SOURCE_FABRIC:
+            xml_text = self._wire_client.fetch_config(ext_conf_uri, self._wire_client.get_header())
+            extensions_config = ExtensionsConfig(xml_text)
+            changed = fabric_changed or is_startup
+            if fast_track_changed:
+                # If FastTrack changed too, then save the artifacts profile because the next time
+                # we retrieve it, we'll receive a 304 because the etag didn't change
+                self._saved_artifacts_profile = artifacts_profile
+        else:
+            extensions_config = artifacts_profile.transform_to_extensions_config()
+            changed = fast_track_changed or is_startup
+            self._saved_artifacts_profile = artifacts_profile
+
+        if changed:
+            if self._pending_mode == GOAL_STATE_SOURCE_FABRIC:
+                self._remove_extensions_if_necessary(extensions_config)
+                self._pending_fabric_incarnation = str(incarnation)
+            else:
+                self._pending_fast_track_seq_no = artifacts_profile.get_sequence_number()
+            logger.info("Handle extensions updates for {0}, Startup={1}", self._last_mode, is_startup)
+
+        return GenericExtensionsConfig(extensions_config, changed, self)
+
+    def _get_fast_track_details(self):
+        artifacts_profile = None
+        fast_track_changed = False
+        
+        if conf.get_extensions_fast_track_enabled():
+            artifacts_profile = self._wire_client.get_artifacts_profile()
+            if artifacts_profile is None and self._saved_artifacts_profile is not None:
+                logger.periodic_info(logger.EVERY_DAY, "Using previously cached artifacts profile")
```

Comment:
A periodic message may not be appropriate here. If the state changes, let's say, from using the cache, to not using it, to not using it again we should get 2 messages.

Reply:
For an example of code dealing with messages that would be logged too frequently but we need to know about when the state changes, see the use of _last_try_update_goal_state_failed in try_update_goal_state. I was planning on refactoring out that code since we need it in a few places.

Reply:
went back and forth on this one a bit, but added the reset_periodic_msg method already discussed